### PR TITLE
feat(iroh-sync): Queries and "views"

### PR DESCRIPTION
**Pull request by Frando**
## Description

Implementation of #1667, extends #1701

With the following changes after discussions in Discord:

* Remove the notion of views, instead embed the info in the query. If the high-level concept of views makes more sense for people from an API perspective, we can restore it in the client api. However under the hood the query details are different per view, this is why here I expose a single `Query` struct
* Add a query builder with a typestate to only allow possible combinations
* Add an index to the redb (fs) document store to make queries that are sorted by key, or that are filtered by key but not by author, efficient. This also is what allows to do queries for the "latest only" entry per key, without allocating the full result set.

In the process I did a refactoring of the redb store to be safer to use. Especially, I moved the `ouroboros` self-referencing stuff into a `ranges` module, and encapsulated in an inner type to keep the self-referencing compilcations scoped. Also did something I had mind for a while: Add some type-safe abstractions around the range bounds constructions that are used when selecting on redb tables. All this turned out quite nice, IMO.

Also contains the changes from #1772 : 
* Renames `get_one` to `get_exact`
* Add flag to `get_exact` whether to include empty entries or not
* Add `get_one` to `iroh::client::Doc` to get a single entry with the same query mechanisms as `get_many`

Open questions and tasks:
* Naming review of the query builder methods
* Integration of the query parameters in the console

## Change checklist

- [x] Self-review.
- [x] Documentation updates if relevant.
- [ ] Tests if relevant.

### CODE REVIEW

**Comment by dignifiedquire** [comment]
I think you need to rebase 😅 lots of conflicts

File: iroh-sync/src/store.rs [hunk 1]
```diff
@@ -143,233 +142,243 @@ pub trait Store: std::fmt::Debug + Clone + Send + Sync + 'static {
...
+    pub fn key(key: impl AsRef<[u8]>) -> QueryBuilder<FlatQuery> {
+        Self::all().key_exact(key)
     }
-}
 
-impl From<std::ops::Range<u64>> for Range {
-    fn from(value: std::ops::Range<u64>) -> Self {
-        Range::Exclusive(value.start, value.end)
+    /// Create a [Query::all] query filtered by a key prefix.
+    pub fn prefix(prefix: impl AsRef<[u8]>) -> QueryBuilder<FlatQuery> {
```
Comment (dignifiedquire) [inline iroh-sync/src/store.rs:305]: maybe use `key_exact` and `key_prefix` for these too?

File: iroh-sync/src/store.rs [hunk 2]
```diff
@@ -143,233 +142,243 @@ pub trait Store: std::fmt::Debug + Clone + Send + Sync + 'static {
...
+}
+
+/// Fields by which the query can be sorted
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub enum SortBy {
+    /// Sort by key.
+    Key,
+    /// Sort by author.
+    #[default]
+    Author,
```
Comment (dignifiedquire) [inline iroh-sync/src/store.rs:349]: this is author and then key, right?

**Comment by Frando** [comment]
> I think you need to rebase 😅 lots of conflicts

yes will do, started hacking away on #1701

File: iroh-sync/src/store/fs.rs [hunk 3]
```diff
@@ -59,10 +64,22 @@ const NAMESPACES_TABLE: TableDefinition<&[u8; 32], &[u8; 32]> =
...
 type RecordsValue<'a> = (u64, &'a [u8; 64], &'a [u8; 64], u64, &'a [u8; 32]);
+
 type RecordsRange<'a> = TableRange<'a, RecordsId<'static>, RecordsValue<'static>>;
 type RecordsTable<'a> = ReadOnlyTable<'a, RecordsId<'static>, RecordsValue<'static>>;
-type DbResult<T> = Result<T, StorageError>;
+type RecordsReader<'a> = TableRangeReader<'a, RecordsId<'static>, RecordsValue<'static>>;
+
+// Records by key
+// Key: (NamespaceId, Key, AuthorId)
+// Value: same as in RECORDS_TABLE
```
Comment (dignifiedquire) [inline iroh-sync/src/store/fs.rs:77]: should we be duplicating the full value here? seems like we could should store a reference to the records table (not sure how to and if that is actually better..)

File: iroh-sync/src/store.rs (hunk 2 above)
Comment (Frando) [reply iroh-sync/src/store.rs]: Yup, renamed for clarity in the latest commit.

File: iroh-sync/src/store.rs (hunk 1 above)
Comment (Frando) [reply iroh-sync/src/store.rs]: done

File: iroh-sync/src/store/fs.rs (hunk 3 above)
Comment (Frando) [reply iroh-sync/src/store/fs.rs]: Yeah, not sure myself. It's one extra lookup vs. 156 bytes (value size) per entry.

In the most recent commit, I changed it to not duplicate the value but instead do the extra lookup.

File: iroh-sync/src/store/fs/migrations.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                    *e = (timestamp, key.to_vec());
+                }
+            })
+            .or_insert_with(|| (timestamp, key.to_vec()));
+    }
+    let len = heads.len();
+    for ((namespace, author), (timestamp, key)) in heads {
+        latest_table.insert((&namespace, &author), (timestamp, key.as_slice()))?;
+    }
+    info!("Migration finished (inserted {len} entries)");
```
Comment (dignifiedquire) [inline iroh-sync/src/store/fs/migrations.rs:36]: maybe print the migration number?

File: iroh-sync/src/store/fs.rs [hunk 5]
```diff
@@ -148,11 +112,15 @@ impl Store {
...
             let mut latest_table = write_tx.open_table(LATEST_TABLE)?;
             let _table = write_tx.open_multimap_table(NAMESPACE_PEERS_TABLE)?;
-            let _table = write_tx.open_table(RECORDS_BY_KEY_TABLE)?;
+            let mut by_key_table = write_tx.open_table(RECORDS_BY_KEY_TABLE)?;
 
             // migration 001: populate latest table if it was empty before
             if latest_table.is_empty()? && !records_table.is_empty()? {
-                migration_001_populate_latest_table(&records_table, &mut latest_table)?;
+                migrations::migration_001_populate_latest_table(&records_table, &mut latest_table)?;
+            }
```
Comment (dignifiedquire) [inline iroh-sync/src/store/fs.rs:129]: we should be committing the transactions between individual migrations

File: iroh-sync/src/store/fs/bounds.rs [hunk 6]
```diff
@@ -0,0 +1,169 @@
+use std::ops::Bound;
+
+use bytes::Bytes;
+
+use crate::{store::KeyFilter, AuthorId, NamespaceId};
+
+use super::{RecordsByKeyId, RecordsByKeyIdOwned, RecordsId, RecordsIdOwned};
+
+/// Bounds on the recors table.
```
Comment (dignifiedquire) [inline iroh-sync/src/store/fs/bounds.rs]: typo

File: iroh-sync/src/store/fs/bounds.rs [hunk 7]
```diff
@@ -0,0 +1,169 @@
+use std::ops::Bound;
+
```
Comment (dignifiedquire) [inline iroh-sync/src/store/fs/bounds.rs:2]: all these bounds are quite important, I wonder if we could write some proptests to make sure they behave as expected

File: iroh-sync/src/store.rs [hunk 8]
```diff
@@ -180,9 +180,10 @@ pub trait Store: std::fmt::Debug + Clone + Send + Sync + 'static {
 #[derive(Debug, Default)]
 pub struct QueryBuilder<K> {
     kind: K,
-    filter_author: AuthorMatcher,
-    filter_key: KeyMatcher,
-    limit_offset: LimitOffset,
+    filter_author: AuthorFilter,
```
Comment (dignifiedquire) [inline iroh-sync/src/store.rs:183]: any specific reasoning for Matcher -> Filter?

File: iroh-sync/src/sync.rs [hunk 9]
```diff
@@ -1943,56 +1943,167 @@ mod tests {
...
+        );
+
+        assert_eq!(
+            query(
+                &store,
+                namespace,
+                Query::all()
+                    .sort_by(SortBy::KeyAuthor, SortDirection::Asc)
+                    .limit(2)
+                    .offset(1)
```
Comment (dignifiedquire) [inline iroh-sync/src/sync.rs:2081]: I think there is no test with key_prefix mixed with offset/limit, that would be good to have

File: iroh-sync/src/store.rs [hunk 10]
```diff
@@ -172,35 +176,250 @@ pub trait Store: std::fmt::Debug + Clone + Send + Sync + 'static {
...
-    Key(Vec<u8>),
-    /// Filter for key prefix
-    Prefix(Vec<u8>),
-    /// Filter by author
-    Author(AuthorId),
-    /// Filter by key prefix and author
-    AuthorAndPrefix(AuthorId, Vec<u8>),
+/// A query builder for document queries.
+#[derive(Debug, Default)]
+pub struct QueryBuilder<K> {
```
Comment (dignifiedquire) [inline iroh-sync/src/store.rs:181]: we should move the whole query stuff into its own file

File: iroh-sync/src/sync.rs [hunk 11]
```diff
@@ -838,6 +838,11 @@ impl RecordIdentifier {
         &self.0
     }
 
+    /// Get this [`RecordIdentifier`] as [Bytes].
+    pub fn to_bytes(&self) -> Bytes {
```
Comment (dignifiedquire) [inline iroh-sync/src/sync.rs:842]: should be `as_bytes` following the convention I believe

File: iroh-sync/src/sync.rs [hunk 12]
```diff
@@ -1903,14 +1916,214 @@ mod tests {
...
+        replica.hash_and_insert("hi/moon", &a2, "a1")?;
+        replica.hash_and_insert("hi", &a3, "a3")?;
+
+        let expect = |rows: &[(&str, &Author)]| {
+            rows.into_iter()
+                .map(|(key, author)| (key.to_string(), author.id()))
+                .collect::<Vec<_>>()
+        };
+
+        assert_eq!(
```
Comment (dignifiedquire) [inline iroh-sync/src/sync.rs:1964]: could write the below a little more readable like this

```rust

struct Test<'a> {
  query: Query,
  res: Vec<(&'static str, &'a AuthorId)>,
}

let tests = [
  Test {  
     query: Query::single_latest_per_key(), 
     res: vec![("hi", &a3), ("hi/moon", &a2), ("hi/world", &a1)],
  },
  // ... 
];

for test in tests {
  assert_eq!(
    query(&store, namespace, test.query),
    expect(&test.res),
    "query: {:#?}", test.query
  );
}
```

File: iroh-sync/src/store/fs/migrations.rs (hunk 4 above)
Comment (Frando) [reply iroh-sync/src/store/fs/migrations.rs]: Fixed in latest commit

File: iroh-sync/src/store/fs.rs (hunk 5 above)
Comment (Frando) [reply iroh-sync/src/store/fs.rs]: Fixed in latest commit

File: iroh-sync/src/store/fs/bounds.rs (hunk 6 above)
Comment (Frando) [reply iroh-sync/src/store/fs/bounds.rs]: Fixed

File: iroh-sync/src/store.rs (hunk 8 above)
Comment (Frando) [reply iroh-sync/src/store.rs:183]: No special reason. Can revert. I think I renamed some variables to `filter` and then thought it to match better with `Filter`. But no preference on my side I think.

File: iroh-sync/src/store/fs/migrations.rs [hunk 13]
```diff
@@ -1,17 +1,49 @@
...
+    Execute(usize),
+}
 
 /// migration 001: populate the latest table (which did not exist before)
-pub fn migration_001_populate_latest_table(
-    records_table: &Table<RecordsId<'static>, RecordsValue<'static>>,
-    latest_table: &mut Table<LatestKey<'static>, LatestValue<'static>>,
-) -> Result<()> {
-    info!("Starting migration: 001_populate_latest_table");
+pub fn migration_001_populate_latest_table(tx: &WriteTransaction) -> Result<MigrateOutcome> {
```
Comment (dignifiedquire) [inline iroh-sync/src/store/fs/migrations.rs:40]: no need for `pub`?

File: iroh-sync/src/sync.rs (hunk 11 above)
Comment (Frando) [reply iroh-sync/src/sync.rs:842]: `as_bytes` already exists and returns `&[u8]`. This `to_bytes` returns [`Bytes`] to skip the extra allocation when recreating a `Bytes`. Open for better ideas on naming, but `as_bytes` is taken..

File: iroh-sync/src/sync.rs (hunk 12 above)
Comment (Frando) [reply iroh-sync/src/sync.rs:1964]: Came up with something similar independently when writing more tests a minute ago ;) but I think I like your's even better, will switch

File: iroh-sync/src/store.rs (hunk 8 above)
Comment (dignifiedquire) [reply iroh-sync/src/store.rs:183]: I think filter is fine, was just curious

File: iroh-sync/src/store.rs (hunk 10 above)
Comment (Frando) [reply iroh-sync/src/store.rs:181]: Done

File: iroh-sync/src/sync.rs (hunk 11 above)
Comment (dignifiedquire) [reply iroh-sync/src/sync.rs:842]: could do 
- `impl AsRef<[u8]>` for the `&[u8]` and
- `fn as_bytes(&self) -> Bytes`

File: iroh-sync/src/store/fs/migrations.rs (hunk 13 above)
Comment (Frando) [reply iroh-sync/src/store/fs/migrations.rs]: Done

File: iroh-sync/src/sync.rs (hunk 12 above)
Comment (dignifiedquire) [reply iroh-sync/src/sync.rs:1964]: :)

File: iroh-sync/src/store.rs (hunk 10 above)
Comment (dignifiedquire) [reply iroh-sync/src/store.rs:181]: I think you only moved the query stuff in the `fs` piece, not here

File: iroh-sync/src/sync.rs (hunk 12 above)
Comment (Frando) [reply iroh-sync/src/sync.rs]: did something inbetween ;)

File: iroh-sync/src/sync.rs (hunk 11 above)
Comment (Frando) [reply iroh-sync/src/sync.rs]: sounds good, changed to that

File: iroh-sync/src/sync.rs (hunk 9 above)
Comment (Frando) [reply iroh-sync/src/sync.rs]: Added a test for that

File: iroh-sync/src/store/fs/bounds.rs (hunk 7 above)
Comment (Frando) [reply iroh-sync/src/store/fs/bounds.rs:2]: One a first thought I can't find a way to write meaningful proptests here. Do you have an idea how they should look or what they should test for?

Comment (dignifiedquire) [reply iroh-sync/src/store/fs/bounds.rs:2]: I don't 😓 mostly thinking out loud

Comment (Frando) [reply iroh-sync/src/store/fs/bounds.rs:2]: I added some manual tests for the edgecases.

**Comment by Frando** [comment]
I merged main and solved the conflicts from #1770.

Also added `sort` and `desc` options to `doc keys` console command.

This is now ready to go from my side.

**Review by dignifiedquire — APPROVED** [review]

**Pull request merged**